### PR TITLE
Add support for tables to `apply` and `applyreduce`

### LINKED ISSUE #90: Handle GeoInterface.jl compatible tables
Probably we should allow GeoInterface.jl compatible tables in `apply` ? (Tables.jl compatible with a `:geometry` column or something else specified by `GeoInterface.geometrycolumns`).

But probably it should also return a Table? to keep the "returns a similar object" pattern.

Could we default to using a `DataFrame` and put DataFrames.jl part in an extension?

This means `applyreduce` could work with just Tables.jl but `apply` would need DataFrames.jl

### PR DESCRIPTION
**Pull request by asinghvi17**
Fixes #90 

This does introduce some type instability, and the table is materialized in column based form.  

A test with the Natural Earth 10m admin-1 dataset reveals:
```julia

julia> @benchmark GO.transform($_scale_by_5, $table)
BenchmarkTools.Trial: 46 samples with 1 evaluation.
 Range (min … max):  102.008 ms … 125.975 ms  ┊ GC (min … max): 0.00% … 14.77%
 Time  (median):     104.848 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   110.125 ms ±   8.571 ms  ┊ GC (mean ± σ):  5.76% ±  7.23%

   ▂▅█▅▅  ▂                                                      
  ▅█████▅██▅▅▁▁▁▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▅█▅███▁▁▅▅▅▁▁▁▅▁▅ ▁
  102 ms           Histogram: frequency by time          126 ms <

 Memory estimate: 90.99 MiB, allocs estimate: 792110.

julia> @benchmark GO.transform($_scale_by_5, $df)
BenchmarkTools.Trial: 52 samples with 1 evaluation.
 Range (min … max):  89.815 ms … 118.400 ms  ┊ GC (min … max): 0.00% … 18.01%
 Time  (median):     92.182 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   97.936 ms ±   8.458 ms  ┊ GC (mean ± σ):  6.45% ±  8.04%

     ▂█▄█                                                       
  ▆▁▄████▄▁▄▆▁▄▄▄▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▁▁▄▆▁▆█▁▆▄█▆ ▁
  89.8 ms         Histogram: frequency by time          110 ms <

 Memory estimate: 87.73 MiB, allocs estimate: 709928.


julia> @benchmark GO.transform($_scale_by_5, $(df.geometry))
BenchmarkTools.Trial: 52 samples with 1 evaluation.
 Range (min … max):  88.676 ms … 112.074 ms  ┊ GC (min … max): 0.00% … 16.77%
 Time  (median):     91.526 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   96.688 ms ±   8.099 ms  ┊ GC (mean ± σ):  6.19% ±  8.02%

      ▃▆▁█                                                      
  ▄▆▁▄████▁▄▆▆▄▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆▆▆▄▇▄▁▄▁▄▁▄▄▄ ▁
  88.7 ms         Histogram: frequency by time          111 ms <

 Memory estimate: 87.27 MiB, allocs estimate: 709503.
```
where `table` is a `Shapefile.Table`, and `df` is `DataFrame(table)`.

Shapefile polygon materialization is not great, so we can probably give it a pass for now.  But the difference between the table and geometry runtimes is highlighted here.  For reference, GI wrapper polygons take 11 ms with this, and GB polygons would probably take even less.

I'll upload a test script for this later.

## TODOs:
- [ ] Tests
- [ ] Docs
- [x] Table support in `applyreduce`
- [ ] Constructing and passing features if FeatureTrait is requested (edge case)

### CODE REVIEW

**Comment by asinghvi17** [comment]
Test script:

```julia

using ArchGDAL, LibGEOS, Shapefile, GeoJSON, WellKnownGeometry, GeometryBasics, GeoInterface

import GeometryOps as GO, GeoInterface as GI

using BenchmarkTools, Chairmarks, CairoMakie, MakieThemes

shp_file = "/Users/anshul/Downloads/ne_10m_admin_0_countries (1)/ne_10m_admin_0_countries.shp"
table = Shapefile.Table(shp_file)

ll2moll = Proj.Transformation("+proj=longlat +datum=WGS84", "+proj=moll")

@benchmark GO.reproject($table, $ll2moll)
@benchmark GO.reproject($(DataFrame(table)), $ll2moll)
@benchmark GO.reproject($(table.geometry), $ll2moll)
@benchmark GO.reproject($(GO.tuples(table.geometry)), $ll2moll)

function _scaleby5(x)
    return x .* 5
end

@b GO.transform($_scaleby5, $table)
@b GO.transform($_scaleby5, $(DataFrame(table)))
@b GO.transform($_scaleby5, $(table.geometry))
@b GO.transform($_scaleby5, $(GO.tuples(table.geometry)))

@b GO.area($(table))
@b GO.area($(DataFrame(table)))
@b GO.area($(table.geometry))
@b GO.area($(GO.tuples(table.geometry)))

using DataFrames
df = DataFrame(table)


```

**Comment by asinghvi17** [comment]
Strange timings:
```julia

julia> @b GO.reproject($table, $ll2moll) seconds=3
112.046 ms (168450 allocs: 15.182 MiB)

julia> @b GO.reproject($(DataFrame(table)), $ll2moll) seconds=3
340.568 ms (86270 allocs: 11.925 MiB)

julia> @b GO.reproject($(go_df), $ll2moll) seconds=3
287.267 ms (72403 allocs: 10.831 MiB)

julia> @b GO.reproject($(table.geometry), $ll2moll) seconds=3
333.583 ms (85846 allocs: 11.463 MiB)

julia> @b GO.reproject($(GO.tuples(table.geometry)), $ll2moll) seconds=3
209.417 ms (71995 allocs: 10.372 MiB)
```
```julia

julia> @b GO.transform($_scaleby5, $table) seconds=3
38.832 ms (154820 allocs: 14.747 MiB)

julia> @b GO.transform($_scaleby5, $(DataFrame(table))) seconds=3
7.468 ms (72636 allocs: 11.490 MiB)

julia> @b GO.transform($_scaleby5, $(go_df)) seconds=3
5.596 ms (59284 allocs: 10.485 MiB)

julia> @b GO.transform($_scaleby5, $(table.geometry)) seconds=3
7.042 ms (72215 allocs: 11.027 MiB)

julia> @b GO.transform($_scaleby5, $(GO.tuples(table.geometry))) seconds=3
4.536 ms (58874 allocs: 10.026 MiB)
```

```julia
julia> @b GO.area($(table)) seconds=3
2.579 ms (5314 allocs: 421.828 KiB)

julia> @b GO.area($(DataFrame(table))) seconds=3
2.622 ms (5315 allocs: 421.844 KiB)

julia> @b GO.area($(go_df)) seconds=3
1.291 ms (16 allocs: 352 bytes)

julia> @b GO.area($(table.geometry)) seconds=3
2.586 ms (5314 allocs: 421.828 KiB)

julia> @b GO.area($(GO.tuples(table.geometry))) seconds=3
1.289 ms (15 allocs: 336 bytes)
```

**Comment by rafaqz** [comment]
Might be the tables implementation in DBFTables.jl:

https://github.com/JuliaData/DBFTables.jl/blob/f22fcf5a50a3b141d6b8845aeab339b7bb1cf1c4/src/DBFTables.jl#L410-L418

Its doing most of the work on `Tables.getcolumn` rather than on construction.

File: src/primitives.jl [hunk 1]
```diff
@@ -212,17 +212,51 @@ end
...
+    geometry_column = first(GI.geometrycolumns(iterable))
+    new_geometry = _apply(f, target, Tables.getcolumn(iterable, geometry_column); threaded, kw...)
+    # Then, we obtain the schema of the table,
+    old_schema = Tables.schema(iterable)
+    # filter the geometry column out,
+    new_names = Iterators.filter(Base.Fix1(!==, geometry_column), old_schema.names)
+    # and try to rebuild the same table as the best type - either the original type of `iterable`,
+    # or a named tuple which is the default fallback.
+    return Tables.materializer(iterable)(
+        merge(
```
Comment (rafaqz) [inline src/primitives.jl:251]: I'm worried about compile time for this when there are 100+ columns. But its ok for now.

File: src/primitives.jl [hunk 2]
```diff
@@ -325,14 +359,35 @@ end
...
+    end
+end
+# In this case, we don't reconstruct the table, but only operate on the geometry column.
+function _applyreduce_to_table(f::F, op::O, target, iterable::IterableType; threaded, init) where {F, O, IterableType}
+    # We extract the geometry column and run `applyreduce` on it.
+    geometry_column = first(GI.geometrycolumns(iterable))
+    return _applyreduce(f, op, target, Tables.getcolumn(iterable, geometry_column); threaded, init)
+end
+# If `applyreduce` wants features, then applyreduce over the rows as `GI.Feature`s.
+___get_col_pair_from_row(row, colname) = colname => Tables.getcolumn(row, colname)
```
Comment (rafaqz) [inline src/primitives.jl:382]:
```suggestion
```

Not used

**Review by rafaqz — APPROVED** [review]
LGTM

File: src/primitives.jl (hunk 1 above)
Comment (asinghvi17) [reply src/primitives.jl:251]: Ah yeah...that might be an issue.

I could use a [Tables.DictColumnTable](https://tables.juliadata.org/stable/#Tables.dictcolumntable) instead, would that work better?

Comment (rafaqz) [reply src/primitives.jl:251]: Yeah that's probably better, its more like a DataFrame

Comment (rafaqz) [reply src/primitives.jl:251]: Allthough that does column-unioning which can be slow for other reasons...

**Comment by skygering** [comment]
I agree we need tests. But it is really readable and easy to understand changes!

File: src/primitives.jl [hunk 3]
```diff
@@ -325,14 +359,34 @@ end
...
+        if threaded isa _True # Try to `collect` and reduce over the vector with threads
+            _applyreduce(f, op, target, collect(iterable); threaded, init)
+        else
+            # Try to `mapreduce` the iterable as-is
+            mapreduce(applyreduce_iterable, op, iterable; init)
+        end
+    end
+end
+# In this case, we don't reconstruct the table, but only operate on the geometry column.
+function _applyreduce_to_table(f::F, op::O, target, iterable::IterableType; threaded, init) where {F, O, IterableType}
```
Comment (rafaqz) [inline src/primitives.jl:376]: I'm not sure it should have `to` in the function name as it doesn't return a table?

File: src/primitives.jl (hunk 1 above)
Comment (asinghvi17) [reply src/primitives.jl:251]: I guess I could manually pass in a Tables.Schema to solve the issue...

**Comment by asinghvi17** [comment]
Latest benchmarks
Screenshot 2024-04-02 at 8 13 38 PM

**Comment by rafaqz** [comment]
Benchmarks look amazing. I think we are at the point were adding a `DBFTables.read(filename, DataFrame)` method like CSV has, and optimizing it, is more important than making anything here faster. 

We will probably find similar things for `GeoJSON.Table` and everything else in JuliaGeo.

**Review by rafaqz — APPROVED** [review]
